include/alp/algorithms/symm_tridiag_eigensolver.hpp

```diff
@@ -0,0 +1,543 @@
+		 *        \a vvt outer product outer(v,v)
+		 *
+		 * @tparam D        Data element type
+		 * @tparam Ring     Type of the semiring used in the computation
+		 * @tparam Minus    Type minus operator used in the computation
+		 * @tparam Divide   Type of divide operator used in the computation
+		 * @param[out] Egvecs    output orthogonal matrix contaning eigenvectors
+		 * @param[out] egvals    output vector containg eigenvalues
+		 * @param[in]  ring A semiring for operations
+		 * @return RC       SUCCESS if the execution was correct
+		 *
+		 */
+		template<
+			typename D,
+			typename VecView1,
+			typename VecImfR1,
+			typename VecImfC1,
+			typename VecView2,
+			typename VecImfR2,
+			typename VecImfC2,
+			typename VecView3,
+			typename VecImfR3,
+			typename VecImfC3,
+			typename OrthogonalType,
+			typename OrthViewType,
+			typename OrthViewImfR,
+			typename OrthViewImfC,
+			class Ring = Semiring< operators::add< D >, operators::mul< D >, identities::zero, identities::one >,
+			class Minus = operators::subtract< D >,
+			class Divide = operators::divide< D >
+		>
+		RC eigensolveDiagPlusOuter(
+			Vector<	D, structures::General,	Dense, VecView1, VecImfR1, VecImfC1 > &egvals,
+			Matrix< D, OrthogonalType, Dense, OrthViewType, OrthViewImfR, OrthViewImfC > &Egvecs,
+			Vector<	D, structures::General, Dense, VecView2, VecImfR2, VecImfC2 > &d,
+			Vector< D, structures::General, Dense, VecView3, VecImfR3, VecImfC3 > &v,
+			const Ring &ring = Ring(),
+			const Minus &minus = Minus(),
+			const Divide &divide = Divide()
+		) {
+			(void)minus;
+			RC rc = SUCCESS;
+
+			const Scalar< D > zero( ring.template getZero< D >() );
+			const Scalar< D > one( ring.template getOne< D >() );
+			const size_t n = nrows( Egvecs );
+			const double eps = 1.e-7;
+
+			// all egvec/val are trivial when the corresponding
+			// element of v is zero
+			size_t count_direct_egvc = 0;
+			size_t count_non_direct_egvc = 0;
+
+			std::vector< size_t > direct_egvc_indx( n, 0 );
+			std::vector< size_t > non_direct_egvc_indx( n, 0 );
+			// the following loop should be replaced by ALP primitives
+			// since v is not sorted it seems that another sort is needed
+			// currently there is no simple way to impement this in ALP
+			for( size_t i = 0; i < n; i++ ) {
+				if( std::abs( v[ i ] ) < eps ) {
+					// in these cases equals are canonical vectors
+					// and eigenvalues are d[i]
+					direct_egvc_indx[ count_direct_egvc ] = i ;
+					++count_direct_egvc;
+				} else {
+					// these cases require complicated egval formula
+					// and for cases where egval is close to the singular point
+					// different algorithm for eigenvectors needs to be implemented
+					non_direct_egvc_indx[ count_non_direct_egvc ] = i;
+					++count_non_direct_egvc;
+				}
+			}
+			direct_egvc_indx.resize( count_direct_egvc );
+			non_direct_egvc_indx.resize( count_non_direct_egvc );
+			alp::Vector< size_t > select_direct( count_direct_egvc );
```

Comment:
Sometimes `alp::` is used, and sometimes it is not. Maybe it would be good to put it everywhere for consistency.